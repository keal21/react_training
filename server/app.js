const {getMovieList, addMovie, editMovie, deleteMovie} = require('./movieList');
const {getFilters} = require('./filters');
const {getSorts} = require('./sorts');

const http = require('http');
const url = require('url');

const hostname = '127.0.0.1';
const port = 4000;

const server = http.createServer((req, res) => {
    res.statusCode = 200;
    res.setHeader('Access-Control-Allow-Origin', "*");
    res.setHeader('Access-Control-Allow-Headers', "*");
    res.setHeader('Content-Type', 'application/json');

    const query = url.parse(req.url, true).query;

    console.log(JSON.stringify(query));

    let movieList;

    // eslint-disable-next-line default-case
    switch (query.action) {
        case 'add':
            addMovie(query);
            break;
        case 'edit':
            editMovie(query);
            break;
        case 'delete':
            deleteMovie(query);
            break;
        case 'get':
            movieList = getMovieList(query.filter, query.sort);
            break;
    }

    const payload = {
        movieList: movieList,
        filters: getFilters(),
        sorts: getSorts(),
    };

    res.end(JSON.stringify(payload));
});

server.listen(port, hostname, () => {
    console.log(`Server running at http://${hostname}:${port}/`);
});
