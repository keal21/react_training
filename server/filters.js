const FILTER_ALL = 'FILTER_ALL';

const FILTERS = [
    {name: 'ALL', value: FILTER_ALL},
    {name: 'DOCUMENTARY', value: 'DOCUMENTARY'},
    {name: 'COMEDY', value: 'COMEDY'},
    {name: 'HORROR', value: 'HORROR'},
    {name: 'CRIME', value: 'CRIME'},
];

function getFilters() {
    return FILTERS;
}

module.exports = {getFilters, FILTER_ALL};
