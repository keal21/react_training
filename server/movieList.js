const {FILTER_ALL} = require('./filters');

let nextMovieId = 0;

let movies = [
    {id: "" + nextMovieId++, title: 'MOVIE 1', overview: 'OVERVIEW 5', releaseDate: '2003', genre: 'DOCUMENTARY', runtime: 'RUNTIME 5', url: 'https://images-na.ssl-images-amazon.com/images/I/51NDT81AY7L._SX343_BO1,204,203,200_.jpg'},
    {id: "" + nextMovieId++, title: 'MOVIE 2', overview: 'OVERVIEW 4', releaseDate: '2002', genre: 'COMEDY', runtime: 'RUNTIME 1', url: 'https://images-na.ssl-images-amazon.com/images/I/51NDT81AY7L._SX343_BO1,204,203,200_.jpg'},
    {id: "" + nextMovieId++, title: 'MOVIE 3', overview: 'OVERVIEW 3', releaseDate: '2004', genre: 'HORROR', runtime: 'RUNTIME 3', url: 'https://images-na.ssl-images-amazon.com/images/I/51NDT81AY7L._SX343_BO1,204,203,200_.jpg'},
    {id: "" + nextMovieId++, title: 'MOVIE 4', overview: 'OVERVIEW 2', releaseDate: '2005', genre: 'CRIME', runtime: 'RUNTIME 2', url: 'https://images-na.ssl-images-amazon.com/images/I/51NDT81AY7L._SX343_BO1,204,203,200_.jpg'},
    {id: "" + nextMovieId++, title: 'MOVIE 5', overview: 'OVERVIEW 1', releaseDate: '2001', genre: 'COMEDY', runtime: 'RUNTIME 4', url: 'https://images-na.ssl-images-amazon.com/images/I/51NDT81AY7L._SX343_BO1,204,203,200_.jpg'},
];

function getMovieList(filter, sort, search) {
    let result = movies.concat();

    if (filter && filter !== FILTER_ALL) {
        result = result.filter(item => item.genre === filter);
    }

    if (search) {
        search = search.toLowerCase();
        result = result.filter(item => item.title.toLowerCase().indexOf(search) !== -1);
    }

    if (sort) {
        result.sort((a, b) => a[sort].localeCompare(b[sort]));
    }

    return result;
}

function addMovie(payload) {
    movies.push({
        id: "" + nextMovieId++,
        title: payload.title,
        overview: payload.overview,
        releaseDate: payload.releaseDate,
        genre: payload.genre,
        runtime: payload.runtime,
        url: payload.url,
    });

    return movies;
}

function editMovie(payload) {
    movies = movies.map(movie =>
        movie.id === payload.id
            ? {
                ...movie,
                id: payload.id,
                title: payload.title,
                overview: payload.overview,
                releaseDate: payload.releaseDate,
                genre: payload.genre,
                runtime: payload.runtime,
                url: payload.url,
            }
            : movie
    );

    return movies;
}

function deleteMovie(payload) {
    movies = movies.filter(movie => movie.id !== payload.id);

    return movies;
}

module.exports = {getMovieList, addMovie, editMovie, deleteMovie};
