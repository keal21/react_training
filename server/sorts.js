const SORTS = [
    {name: 'TITLE', value: 'title'},
    {name: 'OVERVIEW', value: 'overview'},
    {name: 'YEAR', value: 'releaseDate'},
    {name: 'GENRE', value: 'genre'},
];

function getSorts() {
    return SORTS;
}

module.exports = {getSorts};
