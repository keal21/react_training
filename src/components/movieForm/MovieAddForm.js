import React from 'react';
import {connect} from "react-redux";
import {addMovieAction} from "../../actions";
import MovieFormBase from "./MovieFormBase";

const MovieAddForm = ({dispatch}) => (
    <MovieFormBase
        title='Add Movie'
        saveHandler={(item) => dispatch(addMovieAction(item))}/>
)

export default connect()(MovieAddForm);
