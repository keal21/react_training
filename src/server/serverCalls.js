import {fetchMovieListSuccessAction, movieListIsOutdatedAction} from "../actions/movieActions";
import {fetchFiltersSuccessAction} from "../actions/filterActions";
import {fetchSortsSuccessAction} from "../actions/sortActions";

function call(action, item, filter, sort, dispatch, updateMovieList) {
    let url = 'http://127.0.0.1:4000?action=' + action;

    if (item) {
        const params = Object.keys(item).map(key => key + '=' + item[key]).join('&');
        url = url + '&' + params;
    }

    if (filter) {
        url = url + '&filter=' + filter;
    }

    if (sort) {
        url = url + '&sort=' + sort;
    }

    fetch(url)
        .then(async function (res) {
            res = await res.json();

            if (res.movieList) {
                dispatch(fetchMovieListSuccessAction(res.movieList));
            }

            if (!action) {
                dispatch(fetchFiltersSuccessAction(res.filters));
                dispatch(fetchSortsSuccessAction(res.sorts));
            }

            if (updateMovieList) {
                dispatch(movieListIsOutdatedAction());
            }
        })
        .catch(error => {
            //dispatch(fetchMovieListErrorAction(error));
            console.log(error);
        });
}

export function getAllData(dispatch) {
    call('', null, null, null, dispatch);
}

export function callGetMovieList(filter, sort, dispatch) {
    call('get', null, filter, sort, dispatch);
}

export function callAddMovie(item, dispatch) {
    call('add', item, null, null, dispatch, true);
}

export function callEditMovie(item, dispatch) {
    call('edit', item, null, null, dispatch, true);
}

export function callDeleteMovie(item, dispatch) {
    call('delete', item, null, null, dispatch, true);
}
