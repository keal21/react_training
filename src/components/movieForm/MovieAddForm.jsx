import React, { useCallback } from 'react';
import { useDispatch } from 'react-redux';
import { useHistory } from 'react-router-dom';
import MovieFormBase from './MovieFormBase';
import { callAddMovie } from '../../server/serverCalls';
import getNewMovie from '../../reducers/movieEditReducer';

const MovieAddForm = () => {
  const dispatch = useDispatch();
  const history = useHistory();

  const saveHandler = useCallback((item) => {
    callAddMovie(item, dispatch);
    history.push('/');
  }, [dispatch, history]);

  return (
    <MovieFormBase
      title="Add Movie"
      source={getNewMovie()}
      saveHandler={saveHandler}
    />
  );
};

export default MovieAddForm;
