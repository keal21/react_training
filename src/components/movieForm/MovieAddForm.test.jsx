import React from 'react';
import { act, render, screen } from '@testing-library/react';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import userEvent from '@testing-library/user-event';
import MovieAddForm from './MovieAddForm';
import { callAddMovie } from '../../server/serverCalls';

jest.mock('../../server/serverCalls');
jest.mock('react-router-dom', () => ({
  useHistory: () => ({
    push: jest.fn(),
  }),
}));

describe('MovieAddForm', () => {
  const mockStore = configureStore();

  test('snapshot', () => {
    const store = mockStore();

    const { asFragment } = render(<Provider store={store}><MovieAddForm /></Provider>);
    expect(asFragment()).toMatchSnapshot();
  });

  test('validate', async () => {
    const store = mockStore();

    render(<Provider store={store}><MovieAddForm /></Provider>);

    await act(async () => {
      userEvent.click(screen.getByText('Submit'));
    });

    expect(screen.getAllByText('Field is required').length).toBe(6);

    expect(callAddMovie).not.toHaveBeenCalled();
  });

  test('call after submit', async () => {
    const store = mockStore();

    render(<Provider store={store}><MovieAddForm /></Provider>);

    await act(async () => {
      await userEvent.type(screen.getByLabelText('Title'), 'test');
      await userEvent.type(screen.getByLabelText('Release Date'), 'test');
      await userEvent.type(screen.getByLabelText('Url'), 'test');
      await userEvent.type(screen.getByLabelText('Genre'), 'test');
      await userEvent.type(screen.getByLabelText('Overview'), 'test');
      await userEvent.type(screen.getByLabelText('Runtime'), 'test');

      userEvent.click(screen.getByText('Submit'));
    });

    expect(callAddMovie).toHaveBeenCalled();
  });

  test('reset', async () => {
    const store = mockStore();

    render(<Provider store={store}><MovieAddForm /></Provider>);

    await act(async () => {
      await userEvent.type(screen.getByLabelText('Title'), 'test');

      userEvent.click(screen.getByText('Reset'));
    });

    expect(screen.getByLabelText('Title').value).toBe('');
  });
});
