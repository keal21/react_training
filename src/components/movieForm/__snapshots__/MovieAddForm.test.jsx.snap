// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieAddForm snapshot 1`] = `
<DocumentFragment>
  <div
    class="popupBackground"
  >
    <div
      class="popup"
    >
      <div
        class="popupContent"
      >
        <div>
          <h3>
            Add Movie
          </h3>
          <form
            action="#"
          >
            <div>
              <label
                for="title"
              >
                Title
                <br />
              </label>
              <input
                id="title"
                label="Title"
                name="title"
                value=""
              />
              <br />
              <br />
              <br />
            </div>
            <div>
              <label
                for="releaseDate"
              >
                Release Date
                <br />
              </label>
              <input
                id="releaseDate"
                label="Release Date"
                name="releaseDate"
                value=""
              />
              <br />
              <br />
              <br />
            </div>
            <div>
              <label
                for="url"
              >
                Url
                <br />
              </label>
              <input
                id="url"
                label="Url"
                name="url"
                value=""
              />
              <br />
              <br />
              <br />
            </div>
            <div>
              <label
                for="genre"
              >
                Genre
                <br />
              </label>
              <input
                id="genre"
                label="Genre"
                name="genre"
                value=""
              />
              <br />
              <br />
              <br />
            </div>
            <div>
              <label
                for="overview"
              >
                Overview
                <br />
              </label>
              <input
                id="overview"
                label="Overview"
                name="overview"
                value=""
              />
              <br />
              <br />
              <br />
            </div>
            <div>
              <label
                for="runtime"
              >
                Runtime
                <br />
              </label>
              <input
                id="runtime"
                label="Runtime"
                name="runtime"
                value=""
              />
              <br />
              <br />
              <br />
            </div>
            <div>
              <button
                type="button"
              >
                Reset
              </button>
              <div
                class="gap100"
              />
              <button
                type="submit"
              >
                Submit
              </button>
              <div
                class="gap20"
              />
              <button
                type="button"
              >
                Cancel
              </button>
            </div>
          </form>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
